### DISCUSSION TITLE
Argument of type 'boolean' is not assignable to parameter of type 'never'.

### DISCUSSION THREAD

**Discussion by Jakobud** [pmndrs/zustand#2742] [Q&A]
I need to create a store where I save state values using dynamic keys in my application. I have a store like this:

```
export interface StoreState {
  values: {
    darkmode: boolean,
  };
  setValue: {
    darkmode: (newValue: boolean) => void,
  };
}

const useStore = create<StoreState>()(
  (set) => ({
      values: {
        darkmode: true,
      },
      setValue: {
        darkmode: (newValue) => set((state) => ({
          values: {
            ...state.values,
            darkmode: newValue,
          }
        })),  
     }
    }),
) 
```

The idea is that I can have a dynamic number of checkbox inputs in my application. When I create those Checkbox components I pass a unique ID to them and then I access the state value for it like this:

```
// id = 'darkmode';
const checkboxValue = useStore((state) => state.values[id]);
const setCheckboxValue = useStore((state) => state.setValue[id]);
```

So `setCheckboxValue` is now the function used to set the state value of 'darkmode'. But when I try to use `setCheckboxValue` I get this Typescript error:

```
setCheckboxValue(true);
```

```
Argument of type 'boolean' is not assignable to parameter of type 'never'.
```

What am I doing wrong here? How is `setCheckboxValue` not knowing what the input value is supposed to be?

**Comment by dai-shi** [comment]
It looks fine with this playground: https://tsplay.dev/NBr7xN

Can you reproduce the problem in the playground?

**Reply by Jakobud** [reply]
Yes I did it here. Sorry I tried to simplify the original code but it appears when I start adding multiple "values" properties with different types it starts having this problem

https://www.typescriptlang.org/play/?jsx=4#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgCIAvAVwGcYMA7AEzIG4Aod4B7KAjZDgDKMaGhFYcCdnDgA3DABsqaGgC5Esrdp269cGVrlpYwZEoAyGAJ4QqMDQCMIERZgat9Ww7LRNgolAA6sBMMAAWGgxUII4mnjqGuByyNGgwAGpKKhrS2sam5opWtvYaABQKymhOLm6MAJRwALwAfPIQoSlafgHQIWGRcJXZNXDRsSZNbR1dSRy4nMgQDHRwtOKBOM0o6JIAPCJiEtit5Q3l5Wkw0+3lefKj6og+8iYwZpY2dg5wMFAqAA0r16gQGEQ0AEYAAzQ4GyXDwuDXLLVXKvAofIolH4VBhoADuqJUt1eqXSlzokluwweuiqKmedL0ADo2VTsCyGaokfpMZ9it8yuNCcS0LydIiyXgGg0JaD+qEIcN8UTRqTdNdKfRsDT7tLZNymQatGyWRy0FynhLdArgkqhqqxTatFLdLhZUipR6GhwuDwTPxBHAAArEMAAFWsYFUL1koQ0AGs0LYCHAjugTmgANoAciNuYAut1FBg4ooNHQoNwAOYLJYrNYAYRIkHxPA0ACVMMgYCyAGJN-ZhiCR6Oqdo7e6hFpwfPvAU4+y5wFwUvl5q5kDWFDhNDIRPOAAeuY9LXaD2Wq3gyD3B+PYtnGwzaG11PPyJ1lqN2dChd9hhXms1xNneh4QEej47M+WxvrqH4Wua6Rir+TD-t0QHwOEjBMG4TaKGYiazucH7MiBYEPqM5T-CS3TJIY6AwFQUAMHA+ytPsAD0rQLBwQA

**Reply by dai-shi** [reply]
It doesn't seem a problem. It's a correct error.
Because `id` can be `editorWidth`, the argument can be `number`. So, you can't pass `true`.

One solution would be to assert the type: https://tsplay.dev/mA6nZW

**Reply by Jakobud** [reply]
Hmm I'm confused. Isn't `state.setValue[id]` referencing the `state.verticalLayout` method which is expecting a `boolean` input?

**Reply by Jakobud** [reply]
Even if I force it hardcoded `state.values['verticalLayout']` it still doesn't realize that it's referencing a method that is expecting a boolean.

**Reply by Jakobud** [reply]
Is there a more straightforward way to use Zustand to store state that can be accessed using dynamic keys like this?

**Reply by dai-shi** [reply]
> Even if I force it hardcoded `state.values['verticalLayout']` it still doesn't realize that it's referencing a method that is expecting a boolean.



Yes, it does: https://tsplay.dev/m3DJkW

**Reply by dai-shi** [reply]
It's not Zustand issue. It's how TS works.

**Reply by Jakobud** [reply]
Yes I know it's not a Zustand issue. I'm just wondering if using Zustand there is a more traditional way of dealing with this type of situation. What is the best approach?

**Reply by dai-shi** [reply]
For TS only approach, we need type assertions. For more general approach, I'd avoid dynamic keys. For Zustand only approach, we can try getting functions without selectors (using `.getState`), but that doesn't solve the problem actually. Overall, I think my first approach seems the best so far.

I wonder if anyone else has suggestions. There must be many various solutions.

**Reply by dai-shi** [reply]
What's your requirement btw?
What should happen if you call `setCheckboxValue(true);` in the case of `id===editorWidth`?

**Reply by Jakobud** [reply]
Another thing I can think of is not using a `setValue` method but rather a different method for each 'type' like `setBooleanValue` `setStringValue`, etc. Not very DRY though.
